### PR TITLE
Support for NemotronH Nano VLM

### PR DESCRIPTION
**Pull request by danielafrimi**
Adds a new multimodal model implementation: `vllm/model_executor/models/nano_nemotron_vl.py`

for online serving do the following: 

1. `vllm serve <hf-card> --runner generate --max-model-len 8192 --trust_remote_code `

2.  text prompt for video/image
```messages=[{
        "role":
        "user",
        "content": [
            {
                "type": "text",
                # "text": "What's in this image?" #  <image> placeholder is in the prompt
                 "text": "What's in this video and whats on the tv?" # <video> placeholder is in the prompt
            },
            {
                "type": "video_url",
                "video_url": {
                    "url": video_url
                },

                # "type": "image_url",
                # "image_url": {
                #     "url": image_url
                # },
            },
        ],
    }],
```

### CODE REVIEW

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 1]
```diff
@@ -0,0 +1,1050 @@
...
+
+class CustomImageEmbeddingInputs(TypedDict):
+    type: Literal["image_embeds"]
+    data: Union[torch.Tensor, list[torch.Tensor]]
+    """ 
+    A tensor of shape `(num_images, total_image_feature_size, hidden_size)`
+    or a list of tensors of shape `(total_image_feature_size, hidden_size)`
+
+    `hidden_size` must match the hidden size of language model backbone.
+    """
```
Comment (DarkLight1337) [inline vllm/model_executor/models/nano_nemotron_vl.py]: Can you name this after the model's name?

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 2]
```diff
@@ -0,0 +1,1050 @@
...
+CustomImageInputs = Union[CustomImagePixelInputs, CustomImageEmbeddingInputs]
+
+
+class SquaredReLU(nn.Module):
+
+    def forward(self, x):
+        return torch.pow(torch.nn.functional.relu(x), 2)
+
+
+class RMSNorm(nn.Module):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/nano_nemotron_vl.py]: Consider replacing this with `vllm.model_executor.layers.layernorm.RMSNorm`

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 3]
```diff
@@ -0,0 +1,1050 @@
...
+    ])
+    return transform
+
+
+def input_conditioner(x, norm_mean, norm_std):
+    y = (x - norm_mean) / norm_std
+    return y
+
+
+def find_closest_aspect_ratio(aspect_ratio, target_ratios, width, height,
```
Comment (DarkLight1337) [inline vllm/model_executor/models/nano_nemotron_vl.py]: Consider directly importing these functions from `nemotron_vl.py` if they are the same

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 4]
```diff
@@ -0,0 +1,1050 @@
...
+        max_num_tiles=max_num,
+        use_thumbnail=use_thumbnail,
+        idx=idx,
+    )
+
+    pixel_values = torch.stack(images)
+    return pixel_values
+
+
+class BaseCustomProcessor(ABC):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/nano_nemotron_vl.py]: Again please name these after the model's name

**Review by DarkLight1337 — COMMENTED** [review]
Thanks for implementing this in vLLM! Some initial comments

File: vllm/model_executor/models/nano_nemotron_vl.py (hunk 1 above)
Comment (danielafrimi) [reply vllm/model_executor/models/nano_nemotron_vl.py]: fixed, thx

File: vllm/model_executor/models/nano_nemotron_vl.py (hunk 2 above)
Comment (danielafrimi) [reply vllm/model_executor/models/nano_nemotron_vl.py]: replaced, thx

File: vllm/model_executor/models/nano_nemotron_vl.py (hunk 3 above)
Comment (danielafrimi) [reply vllm/model_executor/models/nano_nemotron_vl.py]: done, thanks

**Comment by danielafrimi** [comment]
@DarkLight1337 Thanks for reviewing!

ping you back

File: vllm/config/__init__.py [hunk 5]
```diff
@@ -1542,8 +1542,10 @@ def get_num_layers_by_block_type(
                        for bc in block_configs[start:end])
         else:
             # Hybrid model Jamba
-            layers_block_type_value = getattr(self.hf_config,
-                                              "layers_block_type", None)
+            layers_block_type_value = getattr(
+                self.hf_config, "layers_block_type", None) or getattr(
+                    self.hf_config.get_text_config(), "layers_block_type",
+                    None)
```
Comment (DarkLight1337) [inline vllm/config/__init__.py]:
```suggestion
            layers_block_type_value = getattr(self.hf_text_config,
                                              "layers_block_type", None)
```

**Review by DarkLight1337 — COMMENTED** [review]
Can you add this model to `tests/models/multimodal/processing/test_common.py` to validate the processing logic?

**Comment by DarkLight1337** [comment]
Also need to add it to the test registry `tests/models/registry.py`

**Comment by danielafrimi** [comment]
Thanks! 

@DarkLight1337 There isn't a HF card for this model yet - we will release it in the upcoming month. 
so for adding the model into` tests/models/registry.py` i dont know exactly the model repo name ( i can add it after the model release).
Same for `tests/models/multimodal/processing/test_common.py` it looks like i need to add the HF model_id. 

what do you think?

File: vllm/config/__init__.py (hunk 5 above)
Comment (danielafrimi) [reply vllm/config/__init__.py]: fixed

**Comment by DarkLight1337** [comment]
Inside `tests/models/registry.py`, you can add an entry with a dummy model name and set `is_available_online=False`, that should skip the test for the model

**Comment by danielafrimi** [comment]
@DarkLight1337 Thanks for the review. 
added the model to registry

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 6]
```diff
@@ -0,0 +1,1431 @@
...
+        assert len(tiles) >= 1
+        frames_tensors.append(tiles[0])
+
+    if len(frames_tensors) == 0:
+        frames_tensors = [torch.zeros(3, input_size, input_size)]
+
+    return torch.stack(frames_tensors)
+
+
+class BaseNanoNemotronVLProcessor(ABC):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/nano_nemotron_vl.py:222]: Let's inherit from InternVL processor to avoid duplicate code

Comment (danielafrimi) [reply vllm/model_executor/models/nano_nemotron_vl.py:222]: @DarkLight1337 `BaseNanoNemotronVLProcessor` and `BaseInternVLProcessor` don’t follow the exact same logic.
For example:

In `NanoNemotron` we don’t use `min/max/max_dynamic_patch`, but we do persist normalization attributes.
Image/video processing is largely similar between the two, but `NanoNemotron` assumes that image/video **placeholders** are in the prompt, which leads to slightly different handling (also the processing itself).

Comment (DarkLight1337) [reply vllm/model_executor/models/nano_nemotron_vl.py:222]: Ok sure, thanks for the explanation

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 7]
```diff
@@ -0,0 +1,1401 @@
...
+        self.downsample_ratio = config.downsample_ratio
+        self.ps_version = config.ps_version
+        self.image_tag_type = config.image_tag_type
+
+        self.language_model = init_vllm_registered_model(
+            vllm_config=vllm_config,
+            hf_config=config.text_config,
+            prefix=maybe_prefix(prefix, "language_model"),
+        )
+        self.vision_model = AutoModel.from_config(config.vision_config,
```
Comment (DarkLight1337) [inline vllm/model_executor/models/nano_nemotron_vl.py:908]: Is the vision model part not implemented in vLLM yet?

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 8]
```diff
@@ -0,0 +1,1401 @@
...
+                                                  trust_remote_code=True)
+        self.vision_model.model._initialize_weights = (
+            self.vision_model.model._init_weights)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/nano_nemotron_vl.py:911]: I think this is only used in HF so we can remove this

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 9]
```diff
@@ -0,0 +1,1401 @@
...
+                "In ps_version 'v1', the height and width have not "
+                "been swapped back, which results in a transposed image.",
+                stacklevel=2,
+            )
+        else:
+            x = x.permute(0, 2, 1, 3).contiguous()
+        return x
+
+    def get_vision_model(self, vision_model):
+        pass
```
Comment (DarkLight1337) [inline vllm/model_executor/models/nano_nemotron_vl.py]: Is this intended to stay?

File: vllm/model_executor/models/nano_nemotron_vl.py [hunk 10]
```diff
@@ -0,0 +1,1395 @@
...
+
+    def copy_inputs_before_cuda_graphs(self, input_buffers, **kwargs):
+        return self.language_model.mamba_cache.copy_inputs_before_cuda_graphs(
+            input_buffers, **kwargs)
+
+    def get_seqlen_agnostic_capture_inputs(self, batch_size: int):
+        return (self.language_model.mamba_cache.
+                get_seqlen_agnostic_capture_inputs(batch_size))
+
+    @classmethod
```
Comment (danielafrimi) [inline vllm/model_executor/models/nano_nemotron_vl.py:1381]: @DarkLight1337 added both class function which is required to calc the mamba page size. however those impl are in the llm model and in the vlm one (`vllm/model_executor/models/config.py`). 

what do you think about this WAR? wondering if there any nice impl for this.

File: vllm/model_executor/models/nano_nemotron_vl.py (hunk 7 above)
Comment (danielafrimi) [reply vllm/model_executor/models/nano_nemotron_vl.py:908]: yes, i have a draft PR which implemented the RADIO model with vLLM native. 
after the current PR will merge ill rebase the below one

[PR](https://github.com/vllm-project/vllm/pull/23753)

File: vllm/model_executor/models/nano_nemotron_vl.py (hunk 8 above)
Comment (danielafrimi) [reply vllm/model_executor/models/nano_nemotron_vl.py:911]: removed, thanks

File: vllm/model_executor/models/nano_nemotron_vl.py (hunk 9 above)
Comment (danielafrimi) [reply vllm/model_executor/models/nano_nemotron_vl.py]: removed, thanks

**Comment by danielafrimi** [comment]
@DarkLight1337 fix the above + add some comments. let me know what you think.

File: vllm/model_executor/models/nano_nemotron_vl.py (hunk 10 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/nano_nemotron_vl.py:1381]: @tdoublep probably has a better understanding of this

Comment (tdoublep) [reply vllm/model_executor/models/nano_nemotron_vl.py:1381]: The implementation here looks good to me. We need these class methods `get_mamba_state_shape_from_config` because in `vllm/model_executor/models/config.py` we do some manipulation of the cache config (e.g., modifying the attention block size) and we need to be able to do this without actually instantiating the model.

**Comment by danielafrimi** [comment]
@DarkLight1337 Can we approve it? or is any other changes needed

**Review by DarkLight1337 — APPROVED** [review]
LGTM then, thanks
